src/main.rs

```diff
@@ -1665,16 +1634,98 @@ fn foreign_packages<'a>(
+        .filter(|package| package.is_third_party(&config.policy))
 }
 
-fn first_party_packages<'a>(
+/// All first-party packages, **without** the audit-as-crates-io policy applied
+/// (because it's used for validating that field's value).
+fn first_party_packages_strict<'a>(
     metadata: &'a Metadata,
     config: &'a ConfigFile,
 ) -> impl Iterator<Item = &'a Package> + 'a {
     // Only analyze things from crates.io (no source = path-dep / workspace-member)
     metadata
         .packages
         .iter()
-        .filter(|package| !package.is_third_party(&config.audit_as_crates_io))
+        .filter(|package| !package.is_third_party(&config.policy))
+}
+
+fn check_audit_as_crates_io(
+    out: &mut dyn Write,
+    cfg: &Config,
+    store: &Store,
+) -> Result<(), VetError> {
+    let cache = Cache::acquire(cfg)?;
+    let mut needs_audit_as_entry = vec![];
+    let mut shouldnt_be_audit_as = vec![];
+
+    'packages: for package in first_party_packages_strict(&cfg.metadata, &store.config) {
+        let audit_policy = store
+            .config
+            .policy
+            .get(&package.name)
+            .and_then(|policy| policy.audit_as_crates_io);
+        if audit_policy == Some(false) {
+            // They've explicitly said this is first-party so we don't care about what's in the registry
+            continue;
+        }
+
+        if let Some(index_entry) = cache.query_package_from_index(&package.name) {
+            for index_version in index_entry.versions() {
+                if let Ok(index_ver) = index_version.version().parse::<cargo_metadata::Version>() {
+                    if index_ver == package.version {
+                        // We found a version of this package in the registry!
+                        if audit_policy == None {
+                            // At this point, having no policy is an error
+                            needs_audit_as_entry.push(package);
+                        }
+                        // Now that we've found a version match, we're done with this package
+                        continue 'packages;
+                    }
+                }
+            }
+        }
+
+        // If we reach this point, then we couldn't find a matching package in the registry,
+        // So any `audit-as-crates-io = true` is an error that should be corrected
+        if audit_policy == Some(true) {
+            shouldnt_be_audit_as.push(package);
+        }
+    }
+
+    // ERRORS: these are all fatal diagnostics, but they are batched
+    if !needs_audit_as_entry.is_empty() {
+        writeln!(
+            out,
+            "error: some first-party packages match published crates.io versions"
+        )?;
+        for package in &needs_audit_as_entry {
+            writeln!(out, "  {}:{}", package.name, package.version)?;
+        }
+        writeln!(out)?;
+        writeln!(
+            out,
+            "All of these must have a policy.*.audit-as-crates-io entry"
```

Comment:
hyper-nit: I think it'd be nice to put the symbol `policy.*.audit-as-crates-io` in backticks or single quotes.